File: problems/3181_maximum_total_reward_using_operations_ii/solution.py
from typing import List


class Solution:
    def maxTotalReward(self, rewardValues: List[int]) -> int:
        rewardValues = sorted(set(rewardValues))
        dp = 1
        for v in rewardValues:
            mask = (1 << v) - 1
            dp |= (dp & mask) << v
        return dp.bit_length() - 1

File: problems/3181_maximum_total_reward_using_operations_ii/Solution.java
import java.util.*;

class Solution {
    public int maxTotalReward(int[] rewardValues) {
        TreeSet<Integer> set = new TreeSet<>();
        for (int v : rewardValues) set.add(v);
        java.math.BigInteger dp = java.math.BigInteger.ONE;
        for (int v : set) {
            java.math.BigInteger mask = java.math.BigInteger.ONE.shiftLeft(v).subtract(java.math.BigInteger.ONE);
            dp = dp.or(dp.and(mask).shiftLeft(v));
        }
        return dp.bitLength() - 1;
    }
}
